### PR TITLE
Adding proxier test that uses "iptables-restore --test" to validate i…

### LINKED ISSUE #36850: Enhanced kube-proxy:iptables testing with iptables-restore --test
## FEATURE REQUEST:
At present the rules generated by kube-proxy (pkg/proxy/iptable/proxier) cannot be / are not validated  for syntactic correctness. 

### Options 1
```
iptables-restore --test
```
Given a rules set, it validates if the rules are correct.
This test does more than just syntactic validation.

1. It expects the kernel modules for tables to be loaded (or available for modprobe)
2. You need root privileges to run --test. 
Given this we can run the validation tests as part of e2e tests inside   docker container.

### Option 2
It is still desirable to have this as part of the unit test. 
Update iptables-restore.c to separate out the parser and then use it for proper unit test validation.

### PR DESCRIPTION
**Pull request by yuval-k**
…ptables rules.

fixes kubernetes/kubernetes/#36850

**What this PR does / why we need it**:
This PR tests iptable rules of the proxier on a live node, to make sure they are syntactically valid.

**Which issue this PR fixes**: 

fixes #36850

**Special notes for your reviewer**:

**Release note**:

```release-note
Adding an e2e test to test the iptables proxier tools
```

### CODE REVIEW

**Comment by yuval-k** [comment]
as far as i know the CLA should be signed.. i'm checking this right now..

**Comment by deads2k** [comment]
@thockin or @eparis maybe

**Comment by yuval-k** [comment]
CLA signed

**Comment by cmluciano** [comment]
@k8s-bot ok to test

**Comment by yuval-k** [comment]
working on this

File: test/e2e/proxier.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+			NamespacedName: makeNSN("ns1", "svc1"),
+			Port:           "p80",
+		}
+
+		svcIP := "10.20.30.41"
+		svcPort := 80
+		svcExternalIPs := "50.60.70.81"
+
+		makeServiceMap(fp,
+			makeTestService(svcPortName.Namespace, svcPortName.Name, func(svc *api.Service) {
```
Comment (dcbw) [inline test/e2e/proxier.go]: Might be nice to see the other Service types (LoadBalancer, NodePort, etc) also get tested.  I don't think that would be a bunch of work, but I could be wrong.

File: pkg/ssh/ssh.go [hunk 2]
```diff
@@ -198,11 +198,22 @@ func (d *timeoutDialer) Dial(network, addr string, config *ssh.ClientConfig) (*s
...
+	var inputreader io.Reader
+	if input != "" {
+		inputreader = strings.NewReader(input)
+	}
+	return runSSHCommand(realTimeoutDialer, cmd, user, host, inputreader, signer, true)
 }
 
 // Internal implementation of runSSHCommand, for testing
-func runSSHCommand(dialer sshDialer, cmd, user, host string, signer ssh.Signer, retry bool) (string, string, int, error) {
+func runSSHCommand(dialer sshDialer, cmd, user, host string, input io.Reader, signer ssh.Signer, retry bool) (string, string, int, error) {
```
Comment (dcbw) [inline pkg/ssh/ssh.go]: You'll need to update pkg/ssh/ssh_test.go as well, this is why the bazel and verify tests are failing:

```
W0720 17:30:18.719] Use --strategy=GoCompile=standalone to disable sandboxing for the failing actions.
I0720 17:30:18.819] pkg/ssh/ssh_test.go:307: not enough arguments in call to runSSHCommand
I0720 17:30:18.820] 	have (*mockSSHDialer, string, string, string, ssh.Signer, bool)
I0720 17:30:18.821] 	want (sshDialer, string, string, string, io.Reader, ssh.Signer, bool)
W0720 17:30:20.296] ____Building complete.
```

**Review by dcbw — CHANGES_REQUESTED** [review]

**Comment by dcbw** [comment]
The verify test failed on the bazel updates as well, either run:

`hack/update-bazel.sh`

or if you do per-project GOPATHs (or just rely on the automatic GOPATH in the kube make scripts) you can:

`docker run -it -v $(pwd):/go/src/k8s.io/kubernetes -w /go/src/k8s.io/kubernetes golang:1.8 /bin/bash -c "hack/update-bazel.sh"`

**Comment by cmluciano** [comment]
@yuval-k Need anymore input from us on this?

**Comment by cmluciano** [comment]
cc @dougbtv would you be interested in carrying this PR forward?

**Comment by dougbtv** [comment]
Absolutely, I'll take a look thanks @cmluciano !

**Comment by dougbtv** [comment]
@cmluciano -- tl;dr -- Where I could use some help is -- how should I submit these changes? Should I create a new pull request with my fork of these changes? I'll happily take any advice here. Thanks!!

The whole story...

I've been able to create some fixes to this PR so that it passes the bazel build (at least `make bazel-build` exits 0 and doesn't otherwise complain) -- you'll see that there's a commit from my fork that references this issue, and available in that reference here.

In short, the basics for the fix were to: 

* Update moved `k8s.io/kubernetes/pkg/util/exec` to `k8s.io/utils/exec/testing`
* Update `ssh_test.go` to add parameter when calling `RunSSHCommand(...)`

Also, I had to rebase the commit to account for other changes elsewhere in the codebase to get the bazel builds to complete (e.g. there were [mismatched checksums when I ran bazel](http://pasteall.org/644069))

**Comment by cmluciano** [comment]
@dougbtv Yep, we can close this PR and refer to your new PR with the fixes. Thanks!

**Comment by cmluciano** [comment]
/close

commits included in #54884
